### PR TITLE
Rename fields in status subresource

### PR DESCRIPTION
**Pull request by stanlyDoge**
### Type of change
- Refactoring

### Description
Adding `http://` scheme as prefix to `service:port`, renaming `*address*` to `url`.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Update/write design documentation in `./design`
- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

### CODE REVIEW

File: api/src/main/java/io/strimzi/api/kafka/model/status/KafkaBridgeStatus.java [hunk 1]
```diff
@@ -19,13 +19,15 @@
...
 @EqualsAndHashCode
 public class KafkaBridgeStatus extends Status {
     private static final long serialVersionUID = 1L;
 
     private String httpAddress;
 
+    private int restApiPort;
+
     @Description("The HTTP address at which external client applications can access the Kafka Bridge.")
     public String getHttpAddress() {
```
Comment (tombentley) [inline api/src/main/java/io/strimzi/api/kafka/model/status/KafkaBridgeStatus.java]: I'm confused about what this is. Is the httpAddress the host part of the address, and the RestApiPort the port? If so then the naming is confusing and we should change `httpAddress` → `restApiHost` or similar.

In any case can you clarify in the `@Description` what these properties are?

File: api/src/main/java/io/strimzi/api/kafka/model/status/KafkaConnectStatus.java [hunk 2]
```diff
@@ -19,13 +19,15 @@
...
 @EqualsAndHashCode
 public class KafkaConnectStatus extends Status {
     private static final long serialVersionUID = 1L;
 
     private String restApiAddress;
 
+    private int restApiPort;
+
     @Description("REST API endpoint for managing and monitoring Kafka Connect connectors.")
     public String getRestApiAddress() {
```
Comment (tombentley) [inline api/src/main/java/io/strimzi/api/kafka/model/status/KafkaConnectStatus.java]: To me `address` means `host+port`, so if what is held in this property was like `foobar:1234`, but is now `foobar` with `1234` in the `restApiAddress` then we should rename this property to `restApiHost`.

And we should use consistent language (consistent with the wording for the bridge etc) in the `@Description` to explain what this is.

**Comment by tombentley** [comment]
> In the KafkaStatus, we have this ListenerAddress class.

The nesting would make life a little more difficult if we later needed to expose additional connection-related info (e.g. exposing a url in addition to host + port), so I'm inclined to say we should keep the separate.

**Comment by scholzj** [comment]
> The nesting would make life a little more difficult if we later needed to expose additional connection-related info (e.g. exposing a url in addition to host + port)

@tombentley Could you please elaborate why?

**Comment by tombentley** [comment]
Just that in order to support, say:

```yaml
restApi:
  host: my-bridge.myproject.svc
  port: 8080
  url: https://my-bridge.myproject.svc:8080
```

for one status but just `host` and `port` for other statuses is a little awkward (but merely awkward). I guess it doesn't really matter though, and using a single class does make maintaining consistency easier.

**Comment by scholzj** [comment]
Well, the question are I guess:
* why do you want to url only somewhere and not everywhere
* why not just inherit from ListenerAddress to create HTTPListenerAddress which will have the URL in addition?

It seems to me that having several closely related fields makes it better organized with the nesting.

**Comment by tombentley** [comment]
Yeah, I guess I don't really mind.

File: api/src/main/java/io/strimzi/api/kafka/model/status/HttpListenerAddress.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        generateBuilderPackage = false,
+        builderPackage = "io.fabric8.kubernetes.api.builder"
+)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@JsonPropertyOrder({ "host", "port", "url" })
+@EqualsAndHashCode(callSuper = true)
+public class HttpListenerAddress extends ListenerAddress {
+    private static final long serialVersionUID = 1L;
+
+    private String url;
```
Comment (tombentley) [inline api/src/main/java/io/strimzi/api/kafka/model/status/HttpListenerAddress.java]: When I talked about adding a URL before I was speaking hypothetically: Do we really want to add it as part of this PR?

**Comment by ppatierno** [comment]
@tombentley from what I see the information that the `url` field is adding more than `host` and `port` is just the `scheme` (http or https). What's its purpose?

**Comment by tombentley** [comment]
> @tombentley from what I see the information that the `url` field is adding more than `host` and `port` is just the `scheme` (http or https). What's its purpose?

Yeah, it would add the scheme, and alleviate a user from having to construct the URL themselves (which they will pretty much necessarily have to do for using REST/HTTP). Do we think this is worthwhile? (I didn't mean to derail the PR).

**Comment by scholzj** [comment]
Well, it is there and it is useful. So why not keep it?

**Comment by ppatierno** [comment]
Tbh I don't understand why splitting and then providing one more field where all the information are there together. It seems the only one to find this thing (maybe the entire PR) useless?

**Comment by ppatierno** [comment]
If this PR is about consistency with listeners status why not have together host and port there instead of splitting here, then having one more field (url) to have them together again because for the bridge host and port aren't enough and you need the scheme as well (and adding a new field named `scheme` sounds orrible to me).

**Comment by tombentley** [comment]
I'm starting to think that rather than making these things consistent with each other, we should instead consider how these things are going to be used in practice. For REST/HTTP in practice people will need a URL rather than separate scheme, host, port. For Kafka bootstrap they will need an address (`host:port`), since that's what you pass to a client config. If people need the constituent parts separately it's not so difficult to parse the address or URL to get them, but it make life simpler for users in the usual case.

**Comment by ppatierno** [comment]
Exactly! So I would have the listeners status with host and port together (address) and bridge status with scheme, host and port together (url) ... I don't find useful in having distinct fields for them.

**Comment by scholzj** [comment]
So should we just close this? Is that what you are suggesting @ppatierno @tombentley?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaBridgeAssemblyOperator.java [hunk 4]
```diff
@@ -108,7 +108,7 @@ public KafkaBridgeAssemblyOperator(Vertx vertx, PlatformFeaturesAvailability pfa
                 if (bridge.getHttp() != null) {
                     port = bridge.getHttp().getPort();
                 }
-                kafkaBridgeStatus.setHttpAddress(bridge.getServiceName() + "." + namespace + ".svc:" + port);
+                kafkaBridgeStatus.setUrl("http://" + bridge.getServiceName() + "." + namespace + ".svc:" + port);
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaBridgeAssemblyOperator.java]: Should this be in KafkaBridgeResources like it is for Connect?

Comment (stanlyDoge) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaBridgeAssemblyOperator.java]: Can be.

File: api/src/main/java/io/strimzi/api/kafka/model/KafkaBridgeResources.java [hunk 5]
```diff
@@ -48,4 +48,14 @@ public static String serviceAccountName(String clusterName) {
         return deploymentName(clusterName);
     }
 
+    /**
+     * Returns the URL the Kafka Bridge for a {@code KafkaBridge} cluster of the given name.
+     * @param clusterName  The {@code metadata.name} of the {@code KafkaBridge} resource.
```
Comment (ppatierno) [inline api/src/main/java/io/strimzi/api/kafka/model/KafkaBridgeResources.java]: so it's not a cluster name right?

File: api/src/main/java/io/strimzi/api/kafka/model/status/KafkaBridgeStatus.java [hunk 6]
```diff
@@ -19,19 +19,19 @@
...
 public class KafkaBridgeStatus extends Status {
     private static final long serialVersionUID = 1L;
 
-    private String httpAddress;
+    private String url;
 
-    @Description("The HTTP address at which external client applications can access the Kafka Bridge.")
-    public String getHttpAddress() {
-        return httpAddress;
+    @Description("The URL address at which external client applications can access the Kafka Bridge.")
```
Comment (ppatierno) [inline api/src/main/java/io/strimzi/api/kafka/model/status/KafkaBridgeStatus.java]: Just "URL" not "URL address"

File: api/src/main/java/io/strimzi/api/kafka/model/KafkaConnectResources.java [hunk 7]
```diff
@@ -48,13 +48,13 @@ public static String metricsAndLogConfigMapName(String clusterName) {
     }
 
     /**
-     * Returns the address the Kafka Connect REST API for a {@code KafkaConnect} cluster of the given name.
+     * Returns the URL the Kafka Connect REST API for a {@code KafkaConnect} cluster of the given name.
```
Comment (tombentley) [inline api/src/main/java/io/strimzi/api/kafka/model/KafkaConnectResources.java]:
```suggestion
     * Returns the URL of the Kafka Connect REST API for a {@code KafkaConnect} cluster of the given name.
```

File: api/src/main/java/io/strimzi/api/kafka/model/KafkaBridgeResources.java [hunk 8]
```diff
@@ -48,4 +48,14 @@ public static String serviceAccountName(String clusterName) {
         return deploymentName(clusterName);
     }
 
+    /**
+     * Returns the URL the Kafka Bridge for a {@code KafkaBridge} cluster of the given name.
```
Comment (tombentley) [inline api/src/main/java/io/strimzi/api/kafka/model/KafkaBridgeResources.java]:
```suggestion
     * Returns the URL of the Kafka Bridge for a {@code KafkaBridge} cluster of the given name.
```

File: api/src/main/java/io/strimzi/api/kafka/model/KafkaBridgeResources.java [hunk 9]
```diff
@@ -48,4 +48,14 @@ public static String serviceAccountName(String clusterName) {
...
+     * @param bridgeName  The {@code metadata.name} of the {@code KafkaBridge} resource.
+     * @param namespace The namespace where {@code KafkaBridge} cluster is running.
```
Comment (tombentley) [inline api/src/main/java/io/strimzi/api/kafka/model/KafkaBridgeResources.java]:
```suggestion
     * @param namespace The namespace where the {@code KafkaBridge} cluster is running.
```

File: api/src/main/java/io/strimzi/api/kafka/model/KafkaConnectResources.java [hunk 10]
```diff
@@ -48,13 +48,13 @@ public static String metricsAndLogConfigMapName(String clusterName) {
...
      * @param clusterName  The {@code metadata.name} of the {@code KafkaConnect} resource.
      * @param namespace The namespace where {@code KafkaConnect} cluster is running.
      * @param port The port on which the {@code KafkaConnect} API is available.
      * @return The REST address of {@code KafkaConnect} API.
```
Comment (tombentley) [inline api/src/main/java/io/strimzi/api/kafka/model/KafkaConnectResources.java]:
```suggestion
     * @return The base URL of the {@code KafkaConnect} REST API.
```

File: api/src/main/java/io/strimzi/api/kafka/model/status/KafkaConnectStatus.java [hunk 11]
```diff
@@ -19,19 +19,19 @@
...
 public class KafkaConnectStatus extends Status {
     private static final long serialVersionUID = 1L;
 
-    private String restApiAddress;
+    private String url;
 
-    @Description("REST API endpoint for managing and monitoring Kafka Connect connectors.")
-    public String getRestApiAddress() {
-        return restApiAddress;
+    @Description("URL of REST API endpoint for managing and monitoring Kafka Connect connectors.")
```
Comment (tombentley) [inline api/src/main/java/io/strimzi/api/kafka/model/status/KafkaConnectStatus.java]:
```suggestion
    @Description("The URL of the REST API endpoint for managing and monitoring Kafka Connect connectors.")
```

**Pull request merged**